### ISSUE TITLE
checkpoint cannot be loaded as expected

### ISSUE THREAD

**Issue by boqian-li** [riccardomarin/NICP#6]
Hi, thanks for your great work! 
I have set up the env with the instruction in readme. And when I run `PYTHONPATH=. python ./src/lvd_templ/evaluation/evaluation_benchmark.py`

I get error:

```
Traceback (most recent call last):
  File "./src/lvd_templ/evaluation/evaluation_benchmark.py", line 280, in main
    run(cfg)
  File "./src/lvd_templ/evaluation/evaluation_benchmark.py", line 120, in run
    module, MD, train_data, cfg_model = get_model(chk)
  File "./src/lvd_templ/evaluation/evaluation_benchmark.py", line 70, in get_model
    module = model._load_model_state(checkpoint=old_checkpoint, metadata=MD).to(device)
  File "/root/anaconda3/envs/nsr/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 204, in _load_model_state
    keys = model.load_state_dict(checkpoint["state_dict"], strict=strict)
  File "/root/anaconda3/envs/nsr/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2215, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightUniversal:
        size mismatch for model.segm_list.0.5.bias: copying a param with shape torch.Size([93]) from checkpoint, the shape in current model is torch.Size([159]).
        size mismatch for model.segm_list.0.5.weight_g: copying a param with shape torch.Size([93, 1, 1]) from checkpoint, the shape in current model is torch.Size([159, 1, 1]).
        size mismatch for model.segm_list.0.5.weight_v: copying a param with shape torch.Size([93, 512, 1]) from checkpoint, the shape in current model is torch.Size([159, 512, 1]).
        size mismatch for model.segm_list.1.5.bias: copying a param with shape torch.Size([117]) from checkpoint, the shape in current model is torch.Size([93]).
        size mismatch for model.segm_list.1.5.weight_g: copying a param with shape torch.Size([117, 1, 1]) from checkpoint, the shape in current model is torch.Size([93, 1, 1]).
        size mismatch for model.segm_list.1.5.weight_v: copying a param with shape torch.Size([117, 512, 1]) from checkpoint, the shape in current model is torch.Size([93, 512, 1]).
        size mismatch for model.segm_list.2.5.bias: copying a param with shape torch.Size([270]) from checkpoint, the shape in current model is torch.Size([102]).
        size mismatch for model.segm_list.2.5.weight_g: copying a param with shape torch.Size([270, 1, 1]) from checkpoint, the shape in current model is torch.Size([102, 1, 1]).
        size mismatch for model.segm_list.2.5.weight_v: copying a param with shape torch.Size([270, 512, 1]) from checkpoint, the shape in current model is torch.Size([102, 512, 1]).
        size mismatch for model.segm_list.3.5.bias: copying a param with shape torch.Size([102]) from checkpoint, the shape in current model is torch.Size([270]).
        size mismatch for model.segm_list.3.5.weight_g: copying a param with shape torch.Size([102, 1, 1]) from checkpoint, the shape in current model is torch.Size([270, 1, 1]).
        size mismatch for model.segm_list.3.5.weight_v: copying a param with shape torch.Size([102, 512, 1]) from checkpoint, the shape in current model is torch.Size([270, 512, 1]).
        size mismatch for model.segm_list.4.5.bias: copying a param with shape torch.Size([162]) from checkpoint, the shape in current model is torch.Size([75]).
        size mismatch for model.segm_list.4.5.weight_g: copying a param with shape torch.Size([162, 1, 1]) from checkpoint, the shape in current model is torch.Size([75, 1, 1]).
        size mismatch for model.segm_list.4.5.weight_v: copying a param with shape torch.Size([162, 512, 1]) from checkpoint, the shape in current model is torch.Size([75, 512, 1]).
        size mismatch for model.segm_list.5.5.bias: copying a param with shape torch.Size([159]) from checkpoint, the shape in current model is torch.Size([114]).
        size mismatch for model.segm_list.5.5.weight_g: copying a param with shape torch.Size([159, 1, 1]) from checkpoint, the shape in current model is torch.Size([114, 1, 1]).
        size mismatch for model.segm_list.5.5.weight_v: copying a param with shape torch.Size([159, 512, 1]) from checkpoint, the shape in current model is torch.Size([114, 512, 1]).
        size mismatch for model.segm_list.6.5.bias: copying a param with shape torch.Size([114]) from checkpoint, the shape in current model is torch.Size([78]).
        size mismatch for model.segm_list.6.5.weight_g: copying a param with shape torch.Size([114, 1, 1]) from checkpoint, the shape in current model is torch.Size([78, 1, 1]).
        size mismatch for model.segm_list.6.5.weight_v: copying a param with shape torch.Size([114, 512, 1]) from checkpoint, the shape in current model is torch.Size([78, 512, 1]).
        size mismatch for model.segm_list.7.5.bias: copying a param with shape torch.Size([75]) from checkpoint, the shape in current model is torch.Size([120]).
        size mismatch for model.segm_list.7.5.weight_g: copying a param with shape torch.Size([75, 1, 1]) from checkpoint, the shape in current model is torch.Size([120, 1, 1]).
        size mismatch for model.segm_list.7.5.weight_v: copying a param with shape torch.Size([75, 512, 1]) from checkpoint, the shape in current model is torch.Size([120, 512, 1]).
        size mismatch for model.segm_list.8.5.bias: copying a param with shape torch.Size([159]) from checkpoint, the shape in current model is torch.Size([141]).
        size mismatch for model.segm_list.8.5.weight_g: copying a param with shape torch.Size([159, 1, 1]) from checkpoint, the shape in current model is torch.Size([141, 1, 1]).
        size mismatch for model.segm_list.8.5.weight_v: copying a param with shape torch.Size([159, 512, 1]) from checkpoint, the shape in current model is torch.Size([141, 512, 1]).
        size mismatch for model.segm_list.9.5.bias: copying a param with shape torch.Size([78]) from checkpoint, the shape in current model is torch.Size([141]).
        size mismatch for model.segm_list.9.5.weight_g: copying a param with shape torch.Size([78, 1, 1]) from checkpoint, the shape in current model is torch.Size([141, 1, 1]).
        size mismatch for model.segm_list.9.5.weight_v: copying a param with shape torch.Size([78, 512, 1]) from checkpoint, the shape in current model is torch.Size([141, 512, 1]).
        size mismatch for model.segm_list.10.5.bias: copying a param with shape torch.Size([120]) from checkpoint, the shape in current model is torch.Size([159]).
        size mismatch for model.segm_list.10.5.weight_g: copying a param with shape torch.Size([120, 1, 1]) from checkpoint, the shape in current model is torch.Size([159, 1, 1]).
        size mismatch for model.segm_list.10.5.weight_v: copying a param with shape torch.Size([120, 512, 1]) from checkpoint, the shape in current model is torch.Size([159, 512, 1]).
        size mismatch for model.segm_list.11.5.bias: copying a param with shape torch.Size([141]) from checkpoint, the shape in current model is torch.Size([114]).
        size mismatch for model.segm_list.11.5.weight_g: copying a param with shape torch.Size([141, 1, 1]) from checkpoint, the shape in current model is torch.Size([114, 1, 1]).
        size mismatch for model.segm_list.11.5.weight_v: copying a param with shape torch.Size([141, 512, 1]) from checkpoint, the shape in current model is torch.Size([114, 512, 1]).
        size mismatch for model.segm_list.12.5.bias: copying a param with shape torch.Size([114]) from checkpoint, the shape in current model is torch.Size([165]).
        size mismatch for model.segm_list.12.5.weight_g: copying a param with shape torch.Size([114, 1, 1]) from checkpoint, the shape in current model is torch.Size([165, 1, 1]).
        size mismatch for model.segm_list.12.5.weight_v: copying a param with shape torch.Size([114, 512, 1]) from checkpoint, the shape in current model is torch.Size([165, 512, 1]).
        size mismatch for model.segm_list.13.5.bias: copying a param with shape torch.Size([141]) from checkpoint, the shape in current model is torch.Size([99]).
        size mismatch for model.segm_list.13.5.weight_g: copying a param with shape torch.Size([141, 1, 1]) from checkpoint, the shape in current model is torch.Size([99, 1, 1]).
        size mismatch for model.segm_list.13.5.weight_v: copying a param with shape torch.Size([141, 512, 1]) from checkpoint, the shape in current model is torch.Size([99, 512, 1]).
        size mismatch for model.segm_list.14.5.bias: copying a param with shape torch.Size([99]) from checkpoint, the shape in current model is torch.Size([123]).
        size mismatch for model.segm_list.14.5.weight_g: copying a param with shape torch.Size([99, 1, 1]) from checkpoint, the shape in current model is torch.Size([123, 1, 1]).
        size mismatch for model.segm_list.14.5.weight_v: copying a param with shape torch.Size([99, 512, 1]) from checkpoint, the shape in current model is torch.Size([123, 512, 1]).
        size mismatch for model.segm_list.15.5.bias: copying a param with shape torch.Size([126]) from checkpoint, the shape in current model is torch.Size([117]).
        size mismatch for model.segm_list.15.5.weight_g: copying a param with shape torch.Size([126, 1, 1]) from checkpoint, the shape in current model is torch.Size([117, 1, 1]).
        size mismatch for model.segm_list.15.5.weight_v: copying a param with shape torch.Size([126, 512, 1]) from checkpoint, the shape in current model is torch.Size([117, 512, 1]).
```

After checking the error info, I think maybe the checkpoint wasn't obtained by passing the params needed to correctly instantiate the model as a keyword dictionary. Maybe other mistakes.

**Comment by riccardomarin** [comment]
Hi @boqian-li ,
Thank you for your interest in our work!

This error looks familiar to me- it should have been caused by a wrong segmentation of the mesh using the spectral clustering algorithm (KMeans). This could come from the wrong version of scikit-learn; could you check what version you have installed?

If this doesn't solve the problem, could you attach a `pip freeze`?

Thanks!

**Comment by boqian-li** [comment]
hi @riccardomarin , thanks for your quick response. Yes, you are right, it was caused by the wrong version of sklearn. I install the version 1.1.2 as told in environment.yaml and it works now. Thanks again!

Btw, I noticed that the training code is not released. Would you like to do it in the future?

**Comment by riccardomarin** [comment]
I'm happy that it worked! :) 

The training code is available under the "train" branch of this repository. It is still a bit messy (I plan to polish it in the next future), but it should work.

**Comment by boqian-li** [comment]
Sounds great! Could you please also provide commands for training? Like what's for evaluation.

**Comment by riccardomarin** [comment]
You can find some commands to install and run the training here:
https://github.com/riccardomarin/NICP/tree/train/training

You can run it after the data preprocessing:
https://github.com/riccardomarin/NICP/tree/train/training/data_preprocessing

I will try to add more details about the main parameters of the configuration files in the future (while some are loosely commented in the parsing function `fetch_kwargs` code: 
https://github.com/riccardomarin/NICP/blob/0882749b569dd422063abd941e70bd6d290f7242/training/src/lvd_templ/pl_modules/pl_module_ifnet.py#L208C9-L208C21
)

**Comment by boqian-li** [comment]
cool! I'll try it. 
Thanks for your response！

**Comment by adityamurugan** [comment]
I am getting a similar error but only some of the segments have a shape mismatch. Here's the error:

    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightUniversal:
        size mismatch for model.segm_list.4.5.bias: copying a param with shape torch.Size([162]) from checkpoint, the shape in current model is torch.Size([165]).
        size mismatch for model.segm_list.4.5.weight_g: copying a param with shape torch.Size([162, 1, 1]) from checkpoint, the shape in current model is torch.Size([165, 1, 1]).
        size mismatch for model.segm_list.4.5.weight_v: copying a param with shape torch.Size([162, 512, 1]) from checkpoint, the shape in current model is torch.Size([165, 512, 1]).
        size mismatch for model.segm_list.11.5.bias: copying a param with shape torch.Size([141]) from checkpoint, the shape in current model is torch.Size([138]).
        size mismatch for model.segm_list.11.5.weight_g: copying a param with shape torch.Size([141, 1, 1]) from checkpoint, the shape in current model is torch.Size([138, 1, 1]).
        size mismatch for model.segm_list.11.5.weight_v: copying a param with shape torch.Size([141, 512, 1]) from checkpoint, the shape in current model is torch.Size([138, 512, 1]).

I checked the sklearn version and it's 1.1.2. What am I doing wrong? Here's my pip freeze:

absl-py==2.1.0
addict==2.4.0
aiohappyeyeballs==2.4.3
aiohttp==3.10.11
aiohttp-retry==2.9.1
aiosignal==1.3.1
altair==4.2.2
amqp==5.3.1
annotated-types==0.7.0
antlr4-python3-runtime==4.9.3
anyio==4.5.2
appdirs==1.4.4
arrow==1.3.0
asttokens==2.4.1
async-timeout==5.0.1
asyncssh==2.18.0
atpublic==5.0
attrs==24.2.0
backcall==0.2.0
backoff==2.2.1
backports.zoneinfo==0.2.1
beautifulsoup4==4.12.3
billiard==4.2.1
blessed==1.20.0
blinker==1.8.2
boto3==1.35.65
botocore==1.35.65
Brotli @ file:///croot/brotli-split_1714483155106/work
cachetools==5.5.0
celery==5.4.0
certifi @ file:///croot/certifi_1725551672989/work/certifi
cffi==1.17.1
charset-normalizer @ file:///croot/charset-normalizer_1721748349566/work
click==8.1.7
click-didyoumean==0.3.1
click-plugins==1.1.1
click-repl==0.3.0
colorama==0.4.6
comm==0.2.2
ConfigArgParse==1.7
configobj==5.0.9
contourpy==1.1.1
croniter==1.4.1
cryptography==43.0.3
cycler==0.12.1
Cython==3.0.11
dash==2.18.2
dash-core-components==2.0.0
dash-html-components==2.0.0
dash-table==5.0.0
dateutils==0.6.12
decorator==5.1.1
deepdiff==7.0.1
dictdiffer==0.9.0
diskcache==5.6.3
distro==1.9.0
docker-pycreds==0.4.0
dpath==2.2.0
dulwich==0.22.4
dvc==3.42.0
dvc-data==3.8.0
dvc-http==2.32.0
dvc-objects==3.0.6
dvc-render==1.0.1
dvc-studio-client==0.20.0
dvc-task==0.40.2
editor==1.6.6
entrypoints==0.4
exceptiongroup==1.2.2
executing==2.1.0
fastapi==0.115.5
fastjsonschema==2.20.0
filelock @ file:///croot/filelock_1700591183607/work
Flask==3.0.3
flatten-dict==0.4.2
flufl.lock==7.1.1
fonttools==4.55.0
frozenlist==1.5.0
fsspec==2024.10.0
funcy==2.0
future==1.0.0
fvcore==0.1.5.post20221221
gitdb==4.0.11
GitPython==3.1.43
gmpy2 @ file:///tmp/build/80754af9/gmpy2_1645455532332/work
google-auth==2.36.0
google-auth-oauthlib==1.0.0
grandalf==0.8
grpcio==1.68.0
gto==1.6.2
h11==0.14.0
h5py==3.11.0
hydra-core==1.2.0
idna @ file:///croot/idna_1714398848350/work
imageio==2.35.1
importlib_metadata==8.5.0
importlib_resources==6.4.5
inquirer==3.4.0
iopath==0.1.10
ipython==8.12.3
ipywidgets==8.1.5
iterative-telemetry==0.0.9
itsdangerous==2.2.0
jedi==0.19.2
Jinja2 @ file:///croot/jinja2_1716993405101/work
jmespath==1.0.1
joblib==1.4.2
jsonlines==4.0.0
jsonschema==4.23.0
jsonschema-specifications==2023.12.1
jupyter_core==5.7.2
jupyterlab_widgets==3.0.13
kiwisolver==1.4.7
kombu==5.4.2
lazy_loader==0.4
lightning==2.0.9.post0
lightning-cloud==0.5.70
lightning-utilities==0.11.9
-e git+https://github.com/riccardomarin/NICP.git@12e0a8047ee935dd6ab960d98c26fb48eb5e563f#egg=lvd_templ
Markdown==3.7
markdown-it-py==3.0.0
MarkupSafe @ file:///croot/markupsafe_1704205993651/work
matplotlib==3.7.5
matplotlib-inline==0.1.7
mdurl==0.1.2
mkl-fft @ file:///croot/mkl_fft_1695058164594/work
mkl-random @ file:///croot/mkl_random_1695059800811/work
mkl-service==2.4.0
mpmath @ file:///croot/mpmath_1690848262763/work
multidict==6.1.0
narwhals==1.14.1
nbformat==5.10.4
nest-asyncio==1.6.0
networkx @ file:///croot/networkx_1690561992265/work
nn-template-core==0.1.1
nncore==0.4.0
numpy @ file:///work/mkl/numpy_and_numpy_base_1682953417311/work
nvidia-cublas-cu11==11.11.3.6
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu11==11.8.87
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu11==11.8.89
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu11==11.8.89
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu11==9.1.0.70
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu11==10.9.0.58
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu11==10.3.0.86
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu11==11.4.1.48
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu11==11.7.5.86
nvidia-cusparse-cu12==12.1.0.106
nvidia-nccl-cu11==2.20.5
nvidia-nccl-cu12==2.20.5
nvidia-nvjitlink-cu12==12.6.77
nvidia-nvtx-cu11==11.8.86
nvidia-nvtx-cu12==12.1.105
oauthlib==3.2.2
omegaconf==2.2.2
open3d==0.18.0
opencv-python==4.6.0.66
ordered-set==4.1.0
orjson==3.10.11
packaging==24.2
pandas==2.0.3
parso==0.8.4
pathspec==0.12.1
pathtools==0.1.2
pexpect==4.9.0
pickleshare==0.7.5
pillow @ file:///croot/pillow_1721059439630/work
pkgutil_resolve_name==1.3.10
platformdirs==3.11.0
plotly==5.24.1
portalocker @ file:///croot/portalocker_1682958299664/work
prompt_toolkit==3.0.48
propcache==0.2.0
protobuf==3.20.3
psutil==6.1.0
ptyprocess==0.7.0
pure_eval==0.2.3
pyarrow==17.0.0
pyasn1==0.6.1
pyasn1_modules==0.4.1
pycparser==2.22
pydantic==2.1.1
pydantic_core==2.4.0
pydeck==0.9.1
pyDeprecate==0.3.1
pydot==3.0.2
pygit2==1.13.3
Pygments==2.18.0
pygtrie==2.5.0
PyJWT==2.9.0
Pympler==1.1
pyparsing==3.1.4
pyquaternion==0.9.9
PySocks @ file:///tmp/build/80754af9/pysocks_1605305779399/work
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-multipart==0.0.17
pytorch-lightning==1.5.10
pytorch3d @ git+https://github.com/facebookresearch/pytorch3d.git@c17e6f947a54f167ffcbb61ecb65602bdd1cfd0e
pytz==2024.2
PyWavelets==1.4.1
PyYAML @ file:///croot/pyyaml_1728657952215/work
readchar==4.2.1
referencing==0.35.1
requests @ file:///croot/requests_1721410876868/work
requests-oauthlib==2.0.0
retrying==1.3.4
rich==13.9.4
robust-laplacian==0.2.4
rpds-py==0.20.1
rsa==4.9
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
runs==1.2.2
s3transfer==0.10.3
scikit-image==0.19.3
scikit-learn==1.1.2
scipy==1.10.1
scmrepo==2.0.4
semver==3.0.2
sentry-sdk==2.18.0
setproctitle==1.3.4
shellingham==1.5.4
shortuuid==1.0.13
shtab==1.7.1
six==1.16.0
smmap==5.0.1
sniffio==1.3.1
soupsieve==2.6
sqltrie==0.11.1
stack-data==0.6.3
starlette==0.41.3
starsessions==1.3.0
stqdm==0.0.5
streamlit==1.22.0
sympy @ file:///croot/sympy_1724938189289/work
tabulate==0.9.0
tenacity==8.5.0
tensorboard==2.14.0
tensorboard-data-server==0.7.2
termcolor==2.4.0
threadpoolctl==3.5.0
tifffile==2023.7.10
toml==0.10.2
tomlkit==0.13.2
toolz==1.0.0
torch==1.13.0
torchaudio==0.13.0+cu116
torchmetrics==1.5.2
torchvision==0.14.0+cu116
tornado==6.4.1
tqdm @ file:///croot/tqdm_1724853939799/work
traitlets==5.14.3
trimesh==4.5.2
triton @ file:///opt/conda/envs/bld/conda-bld/torchtriton_1721056825793/work/python
typer==0.13.1
types-python-dateutil==2.9.0.20241003
typing_extensions @ file:///croot/typing_extensions_1715268824938/work
tzdata==2024.2
tzlocal==5.2
urllib3==1.26.20
uvicorn==0.32.0
validators==0.34.0
vine==5.1.0
voluptuous==0.14.2
wandb==0.15.2
watchdog==4.0.2
wcwidth==0.2.13
websocket-client==1.8.0
websockets==12.0
Werkzeug==3.0.6
widgetsnbextension==4.0.13
xmod==1.8.1
yacs==0.1.8
yarl==1.15.2
zc.lockfile==3.0.post1
zipp==3.20.2

**Comment by riccardomarin** [comment]
Hi @adityamurugan ,
From your pip freeze, I can see that not all the library versions match the expectations. For example, your `open3d` has version `0.18.0`, while it should be` 0.15.2`, as stated in the installation script:
https://github.com/riccardomarin/NICP/blob/12e0a8047ee935dd6ab960d98c26fb48eb5e563f/install.sh#L51

open3d is used to reduce the number of template vertices, and different results of this operation may also lead to different segmentations.

Hence I would recommend:
1) Create a new environment
2) Lunch every command provided in installation scripts individually and verify they are successful, and the effect is actually applied to your environment.

As a reference, you can find here the correct segmentation produced by a successful installation (you may also load this one, but this assumes that your template reduction of vertices is also consistent):
https://drive.google.com/file/d/1FR4kY_8C6WaPTI4gBIm868LfdWIw1lHO/view?usp=sharing

**Comment by adityamurugan** [comment]
Yup, that did it. Thanks for the help! Works great.

**Comment by kotakhushi7** [comment]
Title: Segmentation/idxs reproducibility file — Google Drive link is dead (segm_list size mismatch)

Hi @riccardomarin, thanks for the great work. I hit the segm_list size-mismatch on the 1ljjfnbx checkpoint (same as #6). sklearn is 1.1.2, but I'm on Windows + Python 3.8, where open3d 0.15.2 isn't available (only 0.15.1), so the simplify_quadric_decimation reduction produces different idxs and the segmentation doesn't match the checkpoint. The Drive link you shared in #6 for the correct segmentation now returns "file does not exist." Could you please re-share the correct segms_16.npy and/or idxs_10.npy (the reproducibility files referenced in the commented code in dataset_AMASS_occ.py)? That would let me bypass the open3d version dependency. Thank you!
